charms/k8s/src/charm.py

```diff
@@ -16,35 +16,143 @@
 """
 
 import logging
+import re
+import shlex
+import subprocess
+from typing import Optional
 
 import ops
+from charms.k8s.v0.k8sd_api_manager import (
+    InvalidResponseError,
+    K8sdAPIManager,
+    K8sdConnectionError,
+    UnixSocketConnectionFactory,
+)
+from charms.operator_libs_linux.v2.snap import SnapCache, SnapError, SnapState
+from ops.model import WaitingStatus
+from pydantic import ValidationError
 
 # Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 
 VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
-K8SD_SNAP_SOCKET = "/var/snap/k8s/common/var/lib/k8sd/control.socket"
+K8SD_SNAP_SOCKET = "/var/snap/k8s/common/var/lib/k8sd/state/control.socket"
 
 
 class K8sCharm(ops.CharmBase):
-    """Charm the service."""
+    """A charm for managing a K8s cluster via the k8s snap."""
 
     def __init__(self, *args):
-        """Construct.
+        """Initialise the K8s charm.
 
         Args:
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
         self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.install, self._on_install)
 
-    def _on_update_status(self, _event: ops.UpdateStatusEvent):
+        factory = UnixSocketConnectionFactory(unix_socket=K8SD_SNAP_SOCKET)
+        self.api_manager = K8sdAPIManager(factory)
+        self.snap_cache = SnapCache()
+
+    def _apply_snap_requirements(self):
+        """Apply necessary snap requirements for the k8s snap.
+
+        This method executes necessary scripts to ensure that the snap
+        meets the network and interface requirements.
+        """
+        self.unit.status = ops.MaintenanceStatus("Applying K8s requirements")
+        commands = [
+            "/snap/k8s/current/k8s/connect-interfaces.sh",
+            "/snap/k8s/current/k8s/network-requirements.sh",
+        ]
+        for c in commands:
+            subprocess.check_call(shlex.split(c))
+
+    def _bootstrap_k8s_snap(self):
+        """Bootstrap the k8s if it's not already bootstrapped."""
+        if not self.api_manager.is_cluster_bootstrapped():
```

Comment:
@neoaggelos I guess calling `k8s bootstrap` on a bootstrapped cluster should just noop and succeed, WDYT?

Reply:
no, bootstrap is not an idempotent action

Reply:
Hm I agree that this is not idempotent, thatswhy I wrote it is a noop. 

But as a user I want to bootstrap a cluster - if the cluster is already bootstrapped, that's great -> the command should succeed because my intent was fullfilled. 

We shouldn't offload that check to the caller, he doesn't know any internals and then needs to guess if it is bootstrapped already (e.g. calling some endpoint like Mateo did). 
However, this call can fail for various reasons and not just because of an unitialized cluster.